modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java

```diff
@@ -218,78 +223,33 @@ private void destroyRocksDb() throws RocksDBException {
     public void close() throws Exception {
         IgniteUtils.shutdownAndAwaitTermination(snapshotExecutor, 10, TimeUnit.SECONDS);
```

Comment:
This is not what you touch in this PR, just a sidenode. The executor uses an unbounded work queue and a bounded number of threads, and snapshotting operations can probably be pretty long. So, if there is an ongoing snapshot, the executor will be stopped forcibly in the midst of a snapshot. Two (possible?) problems:

1. A snapshot might not be finished?
2. Some tasks submitted to the executor could be enqueued in its work queue and have no chance to be executed, so they will be thrown on the floor due to stop. This will make the corresponding futures hang forever.

Should we file tickets about these problems?

Reply:
Let's imagine the scenario when this method is actually closed: we are stopping an Ignite node. In order to stop the node, we first need to stop the Meta Storage raft group. In which case it's the enclosing component's job to verify that all necessary futures have been resolved before allowing to close itself. What do you think?

Reply:
So, in our case, this would mean that `MetaStorageManager` (the class that now stops `RocksDbKeyValueStorage`) would need to make sure that all the futures are resolved? Currently, its `stop()` method does not seem to bother about it explicitly, but maybe I'm missing something here.

If we push this responsibility to the enclosing component, then this must be a part of the contract, so the contract grows a bit more, and it becomes a little bit less explicit (as the job to deal with such hanging futures is not encapsulated close to the executor). I'm not sure if this is bad or not, to be honest.

Also, we still need this to be done at the enclosing component side, and it does not seem that it's done there yet (again, maybe I just don't see it). Not in this PR, of course.

Reply:
> So, in our case, this would mean that MetaStorageManager (the class that now stops RocksDbKeyValueStorage) would need to make sure that all the futures are resolved? 

yes

> Currently, its stop() method does not seem to bother about it explicitly, but maybe I'm missing something here.

because it's not entirely correct =)

I think I will create a ticket to properly implement the component's stop